lib/bgs/service.rb

```diff
@@ -0,0 +1,197 @@
+        first_nm: payload['first'],
+        middle_nm: payload['middle'],
+        last_nm: payload['last'],
+        suffix_nm: payload['suffix'],
+        brthdy_dt: format_date(payload['birth_date']),
+        birth_state_cd: payload['place_of_birth_state'],
+        birth_city_nm: payload['place_of_birth_city'],
+        file_nbr: payload['va_file_number'],
+        ssn_nbr: payload['ssn'],
+        death_dt: format_date(payload['death_date']),
+        ever_maried_ind: payload['ever_married_ind'],
+        vet_ind: payload['vet_ind'],
+        martl_status_type_cd: 'Married'
+      }.merge(bgs_auth)
+    end
+
+    def with_multiple_attempts_enabled
+      attempt ||= 0
+      yield
+    rescue => e
+      attempt += 1
+      if attempt < MAX_ATTEMPTS
+        notify_of_service_exception(e, __method__.to_s, attempt, :warn)
+        retry
+      end
+
+      notify_of_service_exception(e, __method__.to_s)
+    end
+
+    def bgs_auth
+      {
+        jrn_dt: Time.current.iso8601,
+        jrn_lctn_id: Settings.bgs.client_station_id,
+        jrn_status_type_cd: 'U',
+        jrn_user_id: Settings.bgs.client_username,
+        jrn_obj_id: Settings.bgs.application,
+        ssn: @user[:ssn] # Just here to make the mocks work
+      }
+    end
+
+    def serialize_dependent_result(
+      participant,
+      participant_relationship_type,
+      family_relationship_type,
+      optional_fields = {}
+    )
+
+      {
+        vnp_participant_id: participant[:vnp_ptcpnt_id],
+        participant_relationship_type_name: participant_relationship_type,
+        family_relationship_type_name: family_relationship_type,
+        begin_date: optional_fields[:begin_date],
+        end_date: optional_fields[:end_date],
+        event_date: optional_fields[:event_date],
+        marriage_state: optional_fields[:marriage_state],
+        marriage_city: optional_fields[:marriage_city],
+        divorce_state: optional_fields[:divorce_state],
+        divorce_city: optional_fields[:divorce_city],
+        marriage_termination_type_code: optional_fields[:marriage_termination_type_code],
+        living_expenses_paid_amount: optional_fields[:living_expenses_paid],
+        type: optional_fields[:type]
+      }
+    end
+
+    def dependent_address(lives_with_vet, alt_address)
+      return @dependents_application.dig('veteran_contact_information', 'veteran_address') if lives_with_vet
+
+      alt_address
+    end
+
+    def generate_address(participant_id, address)
+      if address['view:lives_on_military_base'] == true
+        address['military_postal_code'] = address.delete('state_code')
+        address['military_post_office_type_code'] = address.delete('city')
+      end
+
+      create_address(@proc_id, participant_id, address)
+    end
+
+    def notify_of_service_exception(error, method, attempt = nil, status = :error)
+      msg = "Unable to #{method}: #{error.message}: try #{attempt} of #{MAX_ATTEMPTS}"
+      context = { icn: @user[:icn] }
```

Comment:
Is this a PII concern?

Reply:
Good eye. My understanding is no. We have implemented this and shipped elsewhere. I asked this question a while back here are resources I was provided. LMK if you still feel it's a concern.

https://github.com/department-of-veterans-affairs/vets-api/pull/2888

https://dsva.slack.com/archives/CBU0KDSB1/p1591299574037900?thread_ts=1591294003.033000&cid=CBU0KDSB1

Reply:
Ok awesome, just caught up on the above thread. Looks like icn doesn't cause any PII concern and is useful for debugging